sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
+
+  override def compare(l: UTF8String, r: UTF8String): Boolean = {
+    val lCollationID = left.dataType.asInstanceOf[StringType].collationId
+    val rCollationID = right.dataType.asInstanceOf[StringType].collationId
+    // TODO: ensure that lCollationID and rCollationID are equal (compatible)
+    val collationID = if (lCollationID > rCollationID) lCollationID else rCollationID
+    val collatorInfo = CollatorFactory.getInfoForId(collationID)
+    val collator = collatorInfo.collator
+    collator match {
+      // UTF8String collation
+      case null =>
+        collationID match {
+          // UCS_BASIC
+          case 0 =>
+            l.contains(r)
+          // UCS_BASIC_LCASE
+          case 1 =>
+            // scalastyle:off caselocale
+            l.toLowerCase.contains(r.toLowerCase)
+          // scalastyle:on caselocale
+          // Other special collations without ICU collator
+          case _ =>
+            var found = false
+            for (i <- 0 to l.numBytes() - r.numBytes() if !found) {
+              if (collatorInfo.comparator.compare(l.substring(i, i + r.numBytes()), r) == 0) {
+                found = true
+              }
+            }
+            found
+        }
+      // String ICU collation
+      case _ =>
+        val lStr: String = l.toString
+        val rStr: String = r.toString
+        var found = false
+        for (i <- 0 to lStr.length - rStr.length if !found) {
+          if (collator.compare(lStr.substring(i, i + rStr.length), rStr) == 0) {
+            found = true
+          }
+        }
+        found
```

Comment:
Note: for all collations that do have a "collator" (collationID >= 2), manual implementation is required. It appears to me that all manual implementations for built-in string functions will essentially boil down to string manipulation and using "collator.compare" to produce the intended function.

While this implementation resembles the corresponding "contains" library implementation in UTF8String.java, this one could probably be more efficient - open to suggestions!

Also note: while it's thoroughly tested already, here's a quick fiddle for the Scala code: https://scastie.scala-lang.org/sC4SOgI8T9a9lp0y1h18fg

Reply:
Update: this code is now Java-only (implemented directly in CollationFactory::Collation) in the new [PR](https://github.com/apache/spark/pull/45216/files), but here's an equivalent Java fiddle for experimental purposes only: https://onecompiler.com/java/42578zbpf